src/osemgrep/targeting/Unit_semgrepignore.ml

```diff
@@ -0,0 +1,219 @@
+             if should_be_selected then (
+               match status with
+               | Not_ignored ->
+                   printf "[OK] %s: not ignored\n" (Git_path.to_string path)
+               | Ignored ->
+                   printf "[FAIL] %s: ignored\n" (Git_path.to_string path);
+                   error := true)
+             else
+               match status with
+               | Not_ignored ->
+                   printf "[FAIL] %s: not ignored\n" (Git_path.to_string path);
+                   error := true
+               | Ignored ->
+                   printf "[OK] %s: ignored\n" (Git_path.to_string path));
+      assert (not !error))
+
+let tests =
+  Testutil.pack_tests "Semgrepignore"
+    [
+      ("list one file", test_list [ file "a" ]);
+      ( "list hierarchy",
+        test_list
+          [
+            file "a";
+            file "b";
+            symlink "c" "a";
+            dir "dir" [ file "d"; symlink "e" "f"; dir "g" [] ];
+          ] );
+      ( "simple gitignore",
+        test_filter
+          [ File (".gitignore", "*.c"); file "hello.c"; file "hello.ml" ]
+          [ ("/hello.ml", true); ("/hello.c", false) ] );
+      ( "relative paths",
+        test_filter
+          [ File (".gitignore", "*.c"); file "hello.c"; file "hello.ml" ]
+          [ ("hello.ml", true); ("hello.c", false) ] );
+      ( "gitignore + semgrepignore",
+        test_filter
+          [
+            File (".gitignore", "*.c");
+            File (".semgrepignore", "!hello.*");
+            file "hello.c";
+            file "hello.ml";
+          ]
+          [ ("/hello.ml", true); ("/hello.c", true); ("/generated.c", false) ]
+      );
+      ( "semgrepignore alone",
+        test_filter
+          [ File (".semgrepignore", "hello.*"); file "hello.c"; file "bye.c" ]
+          [ ("/hello.c", false); ("/bye.c", true) ] );
+      ( "deep gitignore",
+        test_filter
+          [ dir "dir" [ File (".gitignore", "a"); file "a" ]; file "a" ]
+          [ ("/a", true); ("/dir/a", false) ] );
+      ( "deep semgrepignore + gitignore",
+        test_filter
+          [
+            File (".gitignore", "a");
+            dir "dir"
+              [ File (".semgrepignore", "b"); file "a"; file "b"; file "c" ];
+            file "a";
+            file "b";
+          ]
+          [
+            ("/a", false);
+            ("/b", true);
+            ("/dir/a", false);
+            ("/dir/b", false);
+            ("/dir/c", true);
+          ] );
+      ( "ignore directories only",
+        test_filter
+          [
+            File (".gitignore", "a/");
+            dir "dir" [ file "a" ];
+            dir "a" [ file "b" ];
+          ]
+          [
+            ("/a/", false);
+            ("/a", true);
+            ("/dir/a", true);
+            ("/dir/a/", false);
+            ("/a/b", false);
+          ] );
+      ( "absolute patterns",
+        test_filter
+          [
+            File (".gitignore", "/a\nb/c");
```

Comment:
what's the difference in pattern between /b/c and b/c?
Looks the same to me in behavior.

Reply:
Yes, they should be the same because the presence of a slash in the middle of `b/c` makes it equivalent to `/b/c` as per the gitignore spec. I don't know why this rule exists.

![image](https://user-images.githubusercontent.com/343265/222094619-b4affea4-a47e-4420-be44-7975886f1a1e.png)